[SPARK-13992] Add support for off-heap caching

JoshRosen opened the pull request:
This patch adds support for caching blocks in the executor processes using direct / off-heap memory.
## User-facing changes

**Updated semantics of `OFF_HEAP` storage level**: In Spark 1.x, the `OFF_HEAP` storage level indicated that an RDD should be cached in Tachyon. Spark 2.x removed the external block store API that Tachyon caching was based on (see #10752 / SPARK-12667), so `OFF_HEAP` became an alias for `MEMORY_ONLY_SER`. As of this patch, `OFF_HEAP` means "serialized and cached in off-heap memory or on disk". Via the `StorageLevel` constructor, `useOffHeap` can be set if `serialized == true` and can be used to construct custom storage levels which support replication.

**Storage UI reporting**: the storage UI will now report whether in-memory blocks are stored on- or off-heap.

**Only supported by UnifiedMemoryManager**: for simplicity, this feature is only supported when the default UnifiedMemoryManager is used; applications which use the legacy memory manager (`spark.memory.useLegacyMode=true`) are not currently able to allocate off-heap storage memory, so using off-heap caching will fail with an error when legacy memory management is enabled. Given that we plan to eventually remove the legacy memory manager, this is not a significant restriction.

**Memory management policies:** the policies for dividing available memory between execution and storage are the same for both on- and off-heap memory. For off-heap memory, the total amount of memory available for use by Spark is controlled by `spark.memory.offHeap.size`, which is an absolute size. Off-heap storage memory obeys `spark.memory.storageFraction` in order to control the amount of unevictable storage memory. For example, if `spark.memory.offHeap.size` is 1 gigabyte and Spark uses the default `storageFraction` of 0.5, then up to 500 megabytes of off-heap cached blocks will be protected from eviction due to execution memory pressure. If necessary, we can split `spark.memory.storageFraction` into separate on- and off-heap configurations, but this doesn't seem necessary now and can be done later without any breaking changes.

**Use of off-heap memory does not imply use of off-heap execution (or vice-versa)**: for now, the settings controlling the use of off-heap execution memory (`spark.memory.offHeap.enabled`) and off-heap caching are completely independent, so Spark SQL can be configured to use off-heap memory for execution while continuing to cache blocks on-heap. If desired, we can change this in a followup patch so that `spark.memory.offHeap.enabled` affect the default storage level for cached SQL tables.
## Internal changes
- Rename `ByteArrayChunkOutputStream` to `ChunkedByteBufferOutputStream`
  - It now returns a `ChunkedByteBuffer` instead of an array of byte arrays.
  - Its constructor now accept an `allocator` function which is called to allocate `ByteBuffer`s. This allows us to control whether it allocates regular ByteBuffers or off-heap DirectByteBuffers.
  - Because block serialization is now performed during the unroll process, a `ChunkedByteBufferOutputStream` which is configured with a `DirectByteBuffer` allocator will use off-heap memory for both unroll and storage memory.
- The `MemoryStore`'s MemoryEntries now tracks whether blocks are stored on- or off-heap.
  - `evictBlocksToFreeSpace()` now accepts a `MemoryMode` parameter so that we don't try to evict off-heap blocks in response to on-heap memory pressure (or vice-versa).
- Make sure that off-heap buffers are properly de-allocated during MemoryStore eviction.
- The JVM limits the total size of allocated direct byte buffers using the `-XX:MaxDirectMemorySize` flag and the default tends to be fairly low (< 512 megabytes in some JVMs). To work around this limitation, this patch adds a custom DirectByteBuffer allocator which ignores this memory limit.

SparkQA commented:
**[Test build #53483 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53483/consoleFull)** for PR 11805 at commit [`feffc20`](https://github.com/apache/spark/commit/feffc20ac02565c309bb34444a49cbbab2e7167e).

SparkQA commented:
**[Test build #53483 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53483/consoleFull)** for PR 11805 at commit [`feffc20`](https://github.com/apache/spark/commit/feffc20ac02565c309bb34444a49cbbab2e7167e).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53483/
Test FAILed.

SparkQA commented:
**[Test build #53819 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53819/consoleFull)** for PR 11805 at commit [`222d80b`](https://github.com/apache/spark/commit/222d80ba8503ec430b363d5db66dd9e0b2854c16).

SparkQA commented:
**[Test build #53819 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53819/consoleFull)** for PR 11805 at commit [`222d80b`](https://github.com/apache/spark/commit/222d80ba8503ec430b363d5db66dd9e0b2854c16).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53819/
Test FAILed.

SparkQA commented:
**[Test build #53825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53825/consoleFull)** for PR 11805 at commit [`bf62983`](https://github.com/apache/spark/commit/bf62983b4138b4845bb75bdc9138e00f2280a273).

SparkQA commented:
**[Test build #53825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53825/consoleFull)** for PR 11805 at commit [`bf62983`](https://github.com/apache/spark/commit/bf62983b4138b4845bb75bdc9138e00f2280a273).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53825/
Test FAILed.

JoshRosen committed:
Off-heap caching changes.

SparkQA commented:
**[Test build #54096 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54096/consoleFull)** for PR 11805 at commit [`5a15de2`](https://github.com/apache/spark/commit/5a15de2bd01e842af50e43a65fa45a3b14fecd33).

SparkQA commented:
**[Test build #54096 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54096/consoleFull)** for PR 11805 at commit [`5a15de2`](https://github.com/apache/spark/commit/5a15de2bd01e842af50e43a65fa45a3b14fecd33).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54096/
Test FAILed.

SparkQA commented:
**[Test build #54125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54125/consoleFull)** for PR 11805 at commit [`3c122af`](https://github.com/apache/spark/commit/3c122afb08cceeaf62793d8233605c849ba09208).

SparkQA commented:
**[Test build #54125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54125/consoleFull)** for PR 11805 at commit [`3c122af`](https://github.com/apache/spark/commit/3c122afb08cceeaf62793d8233605c849ba09208).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54125/
Test FAILed.

JoshRosen committed:
Update HistoryServer tests to reflect new StorageLevel.toString

JoshRosen committed:
Fix problems in ChunkedByteBufferOutputStream tests.

JoshRosen committed:
Fix missing classTag in deserialization.

SparkQA commented:
**[Test build #54215 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54215/consoleFull)** for PR 11805 at commit [`334b727`](https://github.com/apache/spark/commit/334b7274e1d12c453fb2f00e4bb543a2902812f5).

JoshRosen committed:
Use UnifiedMemoryManager in BlockManager test suites.

SparkQA commented:
**[Test build #54215 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54215/consoleFull)** for PR 11805 at commit [`334b727`](https://github.com/apache/spark/commit/334b7274e1d12c453fb2f00e4bb543a2902812f5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54215/
Test FAILed.

JoshRosen commented:
/cc @andrewor14, @nongli, @rxin, can you glance over the PR description to check whether the user-facing semantics are okay for now? We might want to change some bits later (maybe to add separate nice shorthands for off-heap-memory-only vs off-heap-memory-and-disk), but I'd like to understand whether any semantics need to change here.

JoshRosen committed:
Discard extra data written by serializer close when discarding partially serialized block.

SparkQA commented:
**[Test build #54239 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54239/consoleFull)** for PR 11805 at commit [`4d9489a`](https://github.com/apache/spark/commit/4d9489a54bc4c2ae9a283a9e2b8bad40d51a2bcd).

rxin commented:
Looks pretty good actually. There are some details about naming and on/off defaults we need to figure out, but that can go in later.

SparkQA commented:
**[Test build #54239 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54239/consoleFull)** for PR 11805 at commit [`4d9489a`](https://github.com/apache/spark/commit/4d9489a54bc4c2ae9a283a9e2b8bad40d51a2bcd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54239/
Test PASSed.

SparkQA commented:
**[Test build #54272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54272/consoleFull)** for PR 11805 at commit [`df8be62`](https://github.com/apache/spark/commit/df8be62b3107a3fe6d01f000721c5e953452cd84).

SparkQA commented:
**[Test build #54272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54272/consoleFull)** for PR 11805 at commit [`df8be62`](https://github.com/apache/spark/commit/df8be62b3107a3fe6d01f000721c5e953452cd84).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54272/
Test PASSed.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into off-heap-caching

JoshRosen commented:
Discovered a potentially major problem: if we bail out of a getOrElseUpdate call and don't end up caching a block and then do not fully consume or dispose of the resulting valuesIterator then we might leak off-heap memory. This problem didn't exist before because we'd never use off-heap memory for unroll space. I may have to add some task completion callbacks to dispose of PartiallySerializedBlocks.

JoshRosen committed:
Tests and several bugfixes.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into off-heap-caching

SparkQA commented:
**[Test build #54388 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54388/consoleFull)** for PR 11805 at commit [`33c9d9b`](https://github.com/apache/spark/commit/33c9d9b70e889d6d2c4131104e5fe954e49bba99).

JoshRosen committed:
Add replication tests and fix more bugs.

SparkQA commented:
**[Test build #54400 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54400/consoleFull)** for PR 11805 at commit [`74427e9`](https://github.com/apache/spark/commit/74427e9352b05f59f5f356b5ce69a4440bd0ddbc).

JoshRosen committed:
Fix accidental early release; dispose on partial read.

SparkQA commented:
**[Test build #54401 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54401/consoleFull)** for PR 11805 at commit [`8a0f944`](https://github.com/apache/spark/commit/8a0f944790b650a05ff9c0d6a6f27d107eeaf69a).

JoshRosen committed:
Use crazy hack to allocate DirectBuffers while ignoring JVM limit.

SparkQA commented:
**[Test build #54404 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54404/consoleFull)** for PR 11805 at commit [`42d0356`](https://github.com/apache/spark/commit/42d0356302be769748af5895f7ce16cba4e0fabe).

JoshRosen commented:
Alright, this should now be ready for an initial review pass. I just pushed a somewhat hacky patch to work around the direct buffer allocation limit (42d0356) and added a few simple tests. I also addressed the potential leak of direct memory in cases where partially serialized blocks' iterators aren't fully consumed.

I'm sure that this probably needs a bit more work and I'll do more self-review tomorrow once I've had time to set this aside for a bit.

JoshRosen commented:
/cc @andrewor14 @davies for review

SparkQA commented:
**[Test build #54388 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54388/consoleFull)** for PR 11805 at commit [`33c9d9b`](https://github.com/apache/spark/commit/33c9d9b70e889d6d2c4131104e5fe954e49bba99).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54388/
Test PASSed.

SparkQA commented:
**[Test build #54404 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54404/consoleFull)** for PR 11805 at commit [`42d0356`](https://github.com/apache/spark/commit/42d0356302be769748af5895f7ce16cba4e0fabe).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54404/
Test FAILed.

SparkQA commented:
**[Test build #54400 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54400/consoleFull)** for PR 11805 at commit [`74427e9`](https://github.com/apache/spark/commit/74427e9352b05f59f5f356b5ce69a4440bd0ddbc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #54401 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54401/consoleFull)** for PR 11805 at commit [`8a0f944`](https://github.com/apache/spark/commit/8a0f944790b650a05ff9c0d6a6f27d107eeaf69a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54401/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54400/
Test PASSed.

davies commented on core/src/main/scala/org/apache/spark/util/io/ChunkedByteBufferOutputStream.scala:
```diff
@@ -67,33 +73,41 @@ class ByteArrayChunkOutputStream(chunkSize: Int) extends OutputStream {
+        StorageUtils.dispose(chunks(lastChunkIndex))
       }
-      ret
+     new ChunkedByteBuffer(ret)
```
indent

davies commented:
Have took a pass on this, looks good to me. Besides unit tests, have you done some stress tests on this (both on heap and off heap)?

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into off-heap-caching

JoshRosen committed:
Indentation fix.

SparkQA commented:
**[Test build #54561 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54561/consoleFull)** for PR 11805 at commit [`a5e9f0d`](https://github.com/apache/spark/commit/a5e9f0d9b85bb1ba65b9ffe2764d3aff1d805166).

SparkQA commented:
**[Test build #54561 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54561/consoleFull)** for PR 11805 at commit [`a5e9f0d`](https://github.com/apache/spark/commit/a5e9f0d9b85bb1ba65b9ffe2764d3aff1d805166).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54561/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:
```diff
@@ -131,8 +129,10 @@ class StorageLevel private(
-    result += (if (useOffHeap) "ExternalBlockStore " else "")
+    if (useMemory) {
+      if (useOffHeap) result += "Memory (off heap) "
+      else result += "Memory (heap) "
```
can you do
```
result += if (useOffHeap) "Memory (off heap)" else "Memory"
```

andrewor14 replied:
probably OK to not say "Memory (heap)" since it doesn't make a ton of sense without the context of off-heap memory

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -416,11 +422,17 @@ private[spark] class BlockManager(
+                _.toInputStream(dispose = false)
+              }.getOrElse {
+                diskBytes.toInputStream(dispose = true)
+              }
```
style nit:
```
val stream = maybeCacheDiskBytesInMemory(info, blockId, level, diskBytes)
  .map { _.toInputStream(dispose = false) }
  .getOrElse { diskBytes.toInputStream(dispose = true) }
```

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:
```diff
@@ -86,7 +83,7 @@ class StorageLevel private(
   }
 
-  def isValid: Boolean = (useMemory || useDisk || useOffHeap) && (replication > 0)
+  def isValid: Boolean = (useMemory || useDisk) && (replication > 0)
```
Do we also want to check that if `offHeap` is true then `useMemory` and `useDisk` must also be true?

JoshRosen replied:
No, because

1. `offHeap` does not necessarily imply `useDisk`, and
2. We already perform well-formedness checks at the start of the constructor (e.g. offHeap == true implies deserialized == false).

JoshRosen replied:
I always thought that `isValid` was somewhat confusingly-named. From its current usage, `isValid` means "if we interpret this StorageLevel as a block's effective StorageLevel (i.e. a descriptor of where it's actually stored) then the block exists in at least one block store." Under that interpretation, I believe that this definition of `isValid` is correct.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -86,7 +91,8 @@ private[spark] class MemoryStore(
   // All accesses of this map are assumed to have manually synchronized on `memoryManager`
-  private val unrollMemoryMap = mutable.HashMap[Long, Long]()
+  private val onHeapUnrollMemoryMap = mutable.HashMap[Long, Long]()
+  private val offHeapUnrollMemoryMap = mutable.HashMap[Long, Long]()
```
can you add a comment here to say the off heap one is only used for bytes? (Is that true)

JoshRosen replied:
Yep, that's true because offHeap implies serialized.

andrewor14 commented on core/src/test/resources/HistoryServerExpectations/one_rdd_storage_json_expectation.json:
```diff
@@ -14,49 +14,49 @@
   }, {
     "blockName" : "rdd_0_7",
-    "storageLevel" : "Memory Deserialized 1x Replicated",
+    "storageLevel" : "Memory (heap) Deserialized 1x Replicated",
```
if you followed my suggestion up there then all these will need to be reverted :)

andrewor14 commented on streaming/src/test/scala/org/apache/spark/streaming/rdd/WriteAheadLogBackedBlockRDDSuite.scala:
```diff
@@ -17,6 +17,7 @@
 package org.apache.spark.streaming.rdd
 
 import java.io.File
+import java.nio.ByteBuffer
```
not used

andrewor14 commented on streaming/src/main/scala/org/apache/spark/streaming/rdd/WriteAheadLogBackedBlockRDD.scala:
```diff
@@ -162,7 +162,8 @@ class WriteAheadLogBackedBlockRDD[T: ClassTag](
       }
-      serializerManager.dataDeserialize(blockId, new ChunkedByteBuffer(dataRead))
+      serializerManager
+        .dataDeserializeStream(blockId, new ChunkedByteBuffer(dataRead).toInputStream())
```
instead of changing these everywhere why not just add an alias for `dataDeserialize` that internally calls the stream version?

JoshRosen replied:
toInputStream() takes an optional parameter for controlling whether buffers are released upon the stream being completely read. Adding an overload kind of masks this option. The one line of extra duplication didn't seem like a big deal to me so I didn't want to add more overloads.

andrewor14 commented on core/src/main/scala/org/apache/spark/util/io/ChunkedByteBufferOutputStream.scala:
```diff
@@ -18,19 +18,25 @@
-class ByteArrayChunkOutputStream(chunkSize: Int) extends OutputStream {
+private[spark] class ChunkedByteBufferOutputStream(
+    chunkSize: Int,
+    allocator: Int => ByteBuffer)
```
seems like this will always be either `ByteBuffer.allocate` or `Platform.allocateDirectBuffer`. It might be better to pass in a boolean here instead `offHeap: Boolean = false` or something and do
```
def allocate(size: Int): ByteBuffer = {
  if (offHeap) Platform.allocateDirectBuffer(size) else ByteBuffer.allocate(size)
}
```
then you don't need to pass a closure around

andrewor14 commented on core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala:
```diff
@@ -117,8 +117,10 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
+   * @param allocator a method for allocating byte buffers
    */
-  def copy(): ChunkedByteBuffer = {
+  def copy(allocator: Int => ByteBuffer): ChunkedByteBuffer = {
```
not used IDT

andrewor14 commented on core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala:
```diff
@@ -117,8 +117,10 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
+  def copy(allocator: Int => ByteBuffer): ChunkedByteBuffer = {
     val copiedChunks = getChunks().map { chunk =>
       // TODO: accept an allocator in this copy method to integrate with mem. accounting systems
       val newChunk = ByteBuffer.allocate(chunk.limit())
```
did you mean to use `allocator` here instead?

JoshRosen replied:
Yes; good catch.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:
```diff
@@ -60,10 +60,7 @@ class StorageLevel private(
   if (useOffHeap) {
-    require(!useDisk, "Off-heap storage level does not support using disk")
-    require(!useMemory, "Off-heap storage level does not support using heap memory")
     require(!deserialized, "Off-heap storage level does not support deserialized storage")
```
as discussed offline, we should also require that `offHeap` implies `useMemory`.

(Additionally `useDisk` should imply `!deserialized`, but we can do that later if you prefer)

JoshRosen replied:
Unfortunately, it's not going to be super easy to do this right now because of how StorageLevel serves the dual purpose of being both a specification of how blocks should be stored and a description of how they're actually stored (i.e. an "effective storage level"). If we added this constraint, the logic for constructing effective storage levels would become much more complicated.

Therefore, I think we should address this issue later and leave the code as-is for now.

andrewor14 commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -144,6 +147,34 @@ public static long reallocateMemory(long address, long oldSize, long newSize) {
 
+  /**
+   * Uses internal JDK APIs to allocate a DirectByteBuffer while ignoring the JVM's
+   * MaxDirectMemorySize limit.
```
can you expand on this briefly to explain why we do this ourselves instead of calling `DirectByteBuffer` directly? It's cause the limit is too low (512m) by default right?

JoshRosen replied:
Yes.

andrewor14 commented:
@JoshRosen Overall looks good. I have not verified comprehensively whether all the things are `disposed` and `released` correctly, but it looks OK from what I can tell. If there are actually leaks I'm sure we'll find them quickly.

andrewor14 commented on core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:
```diff
@@ -74,10 +74,12 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
     val transfer = transferService
       .getOrElse(new NettyBlockTransferService(conf, securityMgr, numCores = 1))
-    val memManager = new StaticMemoryManager(conf, Long.MaxValue, maxMem, numCores = 1)
+    val memManager = UnifiedMemoryManager(conf, numCores = 1)
```
great!

andrewor14 commented:
Forgot to mention, you'll need to update documentation for `spark.memory.storageFraction`. Right now its description is pretty specific to the on-heap stuff.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into off-heap-caching

JoshRosen committed:
Simplify StorageLevel.toString().

JoshRosen committed:
Remove unused import.

JoshRosen committed:
Remove old TODO.

JoshRosen committed:
Fix style nit.

JoshRosen committed:
Expand Platform.allocateDirectBuffer comment.

JoshRosen committed:
Add clarifying comment to offHeapUnrollMemoryMap.

JoshRosen committed:
Use allocator in ChunkedByteBuffer.copy().

SparkQA commented:
**[Test build #54678 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54678/consoleFull)** for PR 11805 at commit [`61920a9`](https://github.com/apache/spark/commit/61920a913ff6fc65d712f12ad565ed52bfb6769e).

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -702,11 +762,12 @@ private[storage] class PartiallySerializedBlock[T](
    */
   def finishWritingToStream(os: OutputStream): Unit = {
     ByteStreams.copy(unrolled.toInputStream(), os)
```
There's actually a bug here: we aren't guaranteed to free the actual offHeap memory here. I think that in my benchmarks it just so happened to be cleaned by the `sun.misc.Cleaner` that I attached to the DirectByteBuffer in my custom allocator. I'm going to take a closer look at this block to try to fix the leak issues here.

JoshRosen committed:
Address possible direct buffer leak.

SparkQA commented:
**[Test build #54678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54678/consoleFull)** for PR 11805 at commit [`61920a9`](https://github.com/apache/spark/commit/61920a913ff6fc65d712f12ad565ed52bfb6769e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54678/
Test FAILed.

SparkQA commented:
**[Test build #54681 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54681/consoleFull)** for PR 11805 at commit [`fde020f`](https://github.com/apache/spark/commit/fde020f32d1d03b99cb58187f65f38e16b88b0fd).

JoshRosen committed:
close() idempotenency fix.

SparkQA commented:
**[Test build #54682 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54682/consoleFull)** for PR 11805 at commit [`a604078`](https://github.com/apache/spark/commit/a604078f339416f60c659ba39609d0ba830c8884).

SparkQA commented:
**[Test build #54681 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54681/consoleFull)** for PR 11805 at commit [`fde020f`](https://github.com/apache/spark/commit/fde020f32d1d03b99cb58187f65f38e16b88b0fd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54681/
Test FAILed.

SparkQA commented:
**[Test build #54682 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54682/consoleFull)** for PR 11805 at commit [`a604078`](https://github.com/apache/spark/commit/a604078f339416f60c659ba39609d0ba830c8884).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54682/
Test PASSed.

andrewor14 commented:
deep-review this please

JoshRosen commented:
Alright, I'm going to merge this into master and will file followup JIRAs to make sure that we update the user-facing documentation and add more tests before 2.0. There are a number of questions to answer about how best to expose the relevant off-heap configurations to users, but I'd rather address them separately.

JoshRosen commented:
Filed https://issues.apache.org/jira/browse/SPARK-14336 as followup
